### PR TITLE
[CuTe,Fwd,Sm100] refactor mla sm100 forward and add page table

### PR DESCRIPTION
**Pull request by jayhshah**
Superseding https://github.com/Dao-AILab/flash-attention/pull/2468.

We refactor the kernel file to remove duplication around splitting V and add page table support.

We also support not providing Q and K (which semantically mean q_pe and k_pe). This computes attention according to the formula
```
O = softmax(scale * (Qv @ V.T)) @ V
```
as in Deepseek v4 core attention.

### CODE REVIEW

File: flash_attn/cute/flash_fwd_sm90.py [hunk 1]
```diff
@@ -557,7 +557,9 @@ def kernel(
                 blocksparse_tensors.cu_total_m_blocks if blocksparse_tensors is not None else None
             ),
             mCuBlockIdxOffsets=(
-                blocksparse_tensors.cu_block_idx_offsets if blocksparse_tensors is not None else None
+                blocksparse_tensors.cu_block_idx_offsets
```
Comment (jayhshah) [inline flash_attn/cute/flash_fwd_sm90.py:560]: This change is just to fix the linter error.

**Comment by jayhshah** [comment]
Note that 64k test in benchmark script throws IMA on cutlass 4.5.0 in the refactor, but this is due to issue identified here: https://github.com/NVIDIA/cutlass/issues/3208

**Comment by jayhshah** [comment]
Benchmark shows ~15% dropoff for paged using cp async loads.

[benchmark_mla_paged.txt](https://github.com/user-attachments/files/27683939/benchmark_mla_paged.txt)

File: flash_attn/cute/interface.py [hunk 2]
```diff
@@ -419,25 +433,33 @@ def _flash_attn_fwd(
...
     qhead_per_kvhead = num_head // num_head_kv
     if pack_gqa is None:
         pack_gqa = qhead_per_kvhead > 1
 
-    is_fp8 = q.dtype in (torch.float8_e4m3fn, torch.float8_e5m2)
-    if is_fp8 and (q.requires_grad or k.requires_grad or v.requires_grad):
+    is_fp8 = v.dtype in (torch.float8_e4m3fn, torch.float8_e5m2)
+    requires_grad = any(t is not None and t.requires_grad for t in [q, k, v, qv])
+    if is_fp8 and requires_grad:
         raise NotImplementedError("FA4 CuTe FP8 backward is not supported yet (forward-only).")
```
Comment (Johnsonms) [inline flash_attn/cute/interface.py:458]: Suggest adding a small interim guard here. The current MLA `qv` path does not support backward yet:

* `FlashAttnFunc.forward` does not `save_for_backward(qv)`
* `_flash_attn_bwd` has no `qv` parameter
* `q.shape[:2]` at line 1384 raises when `q is None` (MLA absorbed path)

The most dangerous case is when `q`, `k`, `v`, and `qv` all require grad: backward runs without crashing but silently computes incorrect gradients as if `qv` were not used in forward.

A lightweight guard converts this into a clear error and also avoids the deep `AttributeError` in the `q=None` path. Since both `FlashAttnFunc` and `FlashAttnVarlenFunc` funnel through `_flash_attn_fwd`, one guard here covers both.

```python
if is_fp8 and requires_grad:
    raise NotImplementedError(
        "FA4 CuTe FP8 backward is not supported yet (forward-only)."
    )

if qv is not None and requires_grad:
    raise NotImplementedError(
        "FA4 CuTe MLA `qv` backward is not implemented yet. "
        "Use `.detach()` on inputs or run under `torch.no_grad()`."
    )
```

Happy to follow up with full MLA backward support if that is in scope for this refactor; otherwise this seems like a cheap and safe first step.

File: flash_attn/cute/interface.py [hunk 3]
```diff
@@ -419,25 +433,33 @@ def _flash_attn_fwd(
...
+    device = v.device
     q_batch_seqlen_shape = (batch_size, seqlen_q) if cu_seqlens_q is None else (total_q,)
-    lse_shape = (batch_size, num_head, seqlen_q) if cu_seqlens_q is None else (num_head, total_q)
-    requires_grad = q.requires_grad or k.requires_grad or v.requires_grad
+
+    if qv is None:
+        lse_shape = (batch_size, num_head, seqlen_q) if cu_seqlens_q is None else (num_head, total_q)
+    else:
+        # num_head contiguous better for MQA in MLA absorbed
+        lse_shape = (batch_size, seqlen_q, num_head) if cu_seqlens_q is None else (total_q, num_head)
```
Comment (Johnsonms) [inline flash_attn/cute/interface.py:467]: The returned `lse` layout depends on `qv`:
- `qv is None`: `(batch, nheads, seqlen_q)` — existing public contract
- `qv is not None`: `(batch, seqlen_q, nheads)` — transposed

Silent contract change for `return_lse=True` callers: not in the public docstrings, and existing tests only check `lse` for non-NaN, so a transposed return goes unnoticed.

Either transpose back at the public API boundary, or document the divergence and add a shape assertion on both paths.

File: benchmarks/benchmark_mla_paged_kv.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+dtype = torch.bfloat16
+seqlen_q = 1
+nheads_q = 128
+nheads_kv = 1  # MQA-128
+headdim = 64
+headdim_v = 512
+causal = True
+
+batch_size = 128
+page_sizes = [None, 16, 64, 128]  # None = non-paged baseline
```
Comment (Johnsonms) [inline benchmarks/benchmark_mla_paged_kv.py:27]: benchmarks/benchmark_mla_paged_kv.py crashes with cudaErrorIllegalAddress  on page_size=16 (cp.async) at seqlen=65536 (total_pages = 128 × 4096 = 524288, V-side indexable ~8 GB). TMA (page_size=128) is clean at the same seqlen — likely 32-bit overflow in byte-offset arithmetic on the cp.async pointer path (cpasync_paged_load_KV @ flash_fwd_mla_sm100.py:1712, compute_X_ptr @ paged_kv.py:158).

**Review by Johnsonms — APPROVED** [review]

File: flash_attn/cute/interface.py (hunk 2 above)
Comment (jayhshah) [reply flash_attn/cute/interface.py:458]: I'll be adding backward in a follow-up PR.

File: benchmarks/benchmark_mla_paged_kv.py (hunk 4 above)
Comment (jayhshah) [reply benchmarks/benchmark_mla_paged_kv.py:27]: This should be fixed in cutlass 4.5.2

File: flash_attn/cute/interface.py (hunk 3 above)
Comment (jayhshah) [reply flash_attn/cute/interface.py:467]: Now documented in docstring for varlen func
